### PR TITLE
Current Sense Changes

### PR DESCRIPTION
**Pull request by badeniran**

### CODE REVIEW

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 1]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
     {
+        resetSenseArray(senses);
         osDelay(SENSE_SETUP_TIME);
+        int ret_val = isSenseStable(senses);
 
-        if (HAL_ADC_PollForConversion(&hadc1, ADC_POLL_TIMEOUT) == HAL_OK)
+        if (ret_val == 1)
+        {
+            HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
+            return senses[0];
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Lets keep just one `return` statement in this function, to keep a hold on how many ways the function can exit.
````
uint32_t current = 0xDEADBEEF
while(...)
{
  ....
  if (ret_val == 1)
  {
    current = senses[0];
    break
  }
.....
}
...
return value
````

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 2]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
+    uint32_t senses[NUM_SENSES];
+    int count = 0;
 
-    for (int counter = 0; counter < 3; counter++)
+    // Want to limit it so we don't get an infinite loop
+    while (count < MAX_CURRENT_READ_ATTEMPTS)
     {
+        resetSenseArray(senses);
         osDelay(SENSE_SETUP_TIME);
+        int ret_val = isSenseStable(senses);
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Does ret_value stand for return value? maybe worth to rename to just `senseStable`

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 3]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
+  Multiple reads are taken and buffered into an array. The maximum and minimum value read are
+  found. Since these values will have the greatest difference, their difference must be less than
+  or equal to the allowable difference. If it is, the CURRENT_SENSE is stable, otherwise, it is deemed
+  not stable
+*/
+int isSenseStable(uint32_t senses[NUM_SENSES])
+{
+    for (int counter = 0; counter < NUM_SENSES; counter++)
+    {
+        for (int i = NUM_SENSES - 1; i > 0; i++)
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Uhhhhh this is an infinite loop, think you mean `i--`?

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 4]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
-            sense = 0xDEADBEEF;
-            break;
+            return -1;
         }
     }
 
-    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
-    return sense;
+    uint32_t max = findMax(senses);
+    uint32_t min = findMin(senses);
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: I think the two funcitons findMax & findMin can be combined, to reduce the amount of looping going on:
````
    for (int i = 1; i < NUM_SENSES; i++)
    {
        uint32_t max = senses[0];
        uint32_t min = senses[0];
        if (senses[i] < min)
        {
            min = senses[i];
        }
  if (senses[] > max)
        {
            max = senses[i];
        }
    }
````

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 5]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
+    return 0xDEADBEEF;
+}
+
+/*
+  Multiple reads are taken and buffered into an array. The maximum and minimum value read are
+  found. Since these values will have the greatest difference, their difference must be less than
+  or equal to the allowable difference. If it is, the CURRENT_SENSE is stable, otherwise, it is deemed
+  not stable
+*/
+int isSenseStable(uint32_t senses[NUM_SENSES])
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: You don't specify the length of an array in a function argument

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 6]
```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
...
+    {
+        return 0;
+    }
+    else
+    {
+        return 1;
+    }
+}
+
+void resetSenseArray(uint32_t senses[NUM_SENSES])
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Don't pass number of items in a function argument.

**Comment by badeniran** [comment]
Lgtm

**Pull request merged**